### PR TITLE
ER-1360 allow revoking recruitment permission

### PR DESCRIPTION
**Pull request by sachpatel**

### CODE REVIEW

File: src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/Dtos/AccountLegalEntityDto.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@ namespace SFA.DAS.ProviderRelationships.Application.Queries.GetAccountProviderLe
     public class AccountLegalEntityDto
     {
         public long Id { get; set; }
+        public string PublicHashedId { get; set; }
```
Comment (elliotritchie) [inline src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/Dtos/AccountLegalEntityDto.cs]: Is this property used?

File: src/SFA.DAS.ProviderRelationships/Services/DasRecruitService.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            _httpClient = recruitApiHttpClientFactory.CreateHttpClient();
+        }
+
+        public async Task<BlockedOrganisationStatus> GetProviderBlockedStatusAsync(long providerUkprn, CancellationToken cancellationToken = default)
+        {
+            var blockedProviderStatusUri = $"/api/providers/{providerUkprn}/status";
+
+            try
+            {
+                var response = await _httpClient.GetAsync(blockedProviderStatusUri, cancellationToken);
```
Comment (elliotritchie) [inline src/SFA.DAS.ProviderRelationships/Services/DasRecruitService.cs]: Injecting and using `IRestHttpClient` might let you avoid writing any boilerplate code here.

File: src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/GetAccountProviderLegalEntityQueryHandler.cs [hunk 3]
```diff
@@ -28,25 +31,33 @@ public async Task<GetAccountProviderLegalEntityQueryResult> Handle(GetAccountPro
...
-            
-            return new GetAccountProviderLegalEntityQueryResult(accountProvider, accountLegalEntity, accountProviderLegalEntity, accountLegalEntitiesCount);
+
+            var providerOrgBlockStatus = await _dasRecruitService.GetProviderBlockedStatusAsync(accountProvider.ProviderUkprn, cancellationToken);
+            var isProviderBlockedFromRecruit = providerOrgBlockStatus != null && providerOrgBlockStatus.Status.Equals(BlockedOrganisationStatusConstants.Blocked);
+
+            return new GetAccountProviderLegalEntityQueryResult(accountProvider, accountLegalEntity, accountProviderLegalEntity, accountLegalEntitiesCount, isProviderBlockedFromRecruit);
+        }
+
+        private Task<VacanciesSummary> GetVacanciesSummary(string employerAccountHashedId, string ukprn, AccountLegalEntityDto accountLegalEntity, CancellationToken cancellationToken)
```
Comment (elliotritchie) [inline src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/GetAccountProviderLegalEntityQueryHandler.cs]: It looks like `AccountHashedId` has been created and set in a bunch of places ultimately to be passed into this method but then it's not called. I'm guessing it will be called eventually as part of another PR?

File: src/SFA.DAS.ProviderRelationships.Messages/Events/UpdatedPermissionsEvent.cs [hunk 4]
```diff
@@ -12,17 +12,26 @@ public class UpdatedPermissionsEvent
...
         public Guid UserRef { get; }
+        public string UserEmailAddress { get; }
+        public string UserFirstName { get; }
+        public string UserLastName { get; }
         public HashSet<Operation> GrantedOperations { get; }
-        public DateTime Updated { get;}
+        public DateTime Updated { get; }
 
-        public UpdatedPermissionsEvent(long accountId, long accountLegalEntityId, long accountProviderId, long accountProviderLegalEntityId, long ukprn, Guid userRef, HashSet<Operation> grantedOperations, DateTime updated)
+        public UpdatedPermissionsEvent(long accountId, long accountLegalEntityId, long accountProviderId,
```
[the remark this answers is not in the document]
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships.Messages/Events/UpdatedPermissionsEvent.cs]: 👍

File: src/SFA.DAS.ProviderRelationships.Web/Controllers/AccountProvidersController.cs [hunk 5]
```diff
@@ -43,6 +43,7 @@ public async Task<ActionResult> Index(AccountProvidersRouteValues routeValues)
             var query = new GetAccountProvidersQuery(routeValues.AccountId.Value);
             var result = await _mediator.Send(query);
             var model = _mapper.Map<AccountProvidersViewModel>(result);
+            model.AccountHashedId = routeValues.AccountHashedId;
```
[the remark this answers is not in the document]
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships.Web/Controllers/AccountProvidersController.cs]: Many links and back links on the pages broke and that is why I have been setting the AccountHashedId onto the ViewModels so in the views I could pass the value in the Url.Action calls.

I know this is messy and goes against the conventions already existing within the codebase. It looks dodgy and is inconsistent. I didn't want to pass the value further into every query class so the handler could map it back into the response objects that then mapped to the ViewModels in the Controllers.

The reason it breaks existing links I believe is because of the RouteValues classes where I have added the property to have the value binded back from the request route.

If I take the changes back out hopefully all the links should work. There is a bit of work I have commented out because it has been descoped for now. At that point are you okay for me to use the Encoding service to encode the AccountId to produce a encodedAccountId I can pass to the Recruit service.

Apologies for the clutter I have introduced.

File: src/SFA.DAS.ProviderRelationships.Web/Controllers/HomeController.cs [hunk 6]
```diff
@@ -16,11 +16,11 @@ public HomeController(IEnvironmentService environmentService, IEmployerUrls empl
         }
 
         [Route]
-        public ActionResult Index()
+        public ActionResult Index(string employerAccountId = null)
```
[the remark this answers is not in the document]
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships.Web/Controllers/HomeController.cs]: 👍

File: src/SFA.DAS.ProviderRelationships.Web/Views/AccountProviderLegalEntities/Get.cshtml [hunk 7]
```diff
@@ -75,71 +82,69 @@
...
-                                                <div class="govuk-radios__item">
-                                                    @if (Model.Operations[i].IsEnabled)
-                                                    {
-                                                        @Html.RadioButtonFor(m => m.Operations[i].IsEnabled, true, new { id = $"operation-{i}-yes", @class = "govuk-radios__input", aria_disabled = "true", disabled = "disabled" })
-                                                    }
-                                                    else
-                                                    {
+                                        @if (Model.IsProviderBlockedFromRecruit)
+                                        {
+                                            <strong>Not allowed</strong>
```
[the remark this answers is not in the document]
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships.Web/Views/AccountProviderLegalEntities/Get.cshtml]: I am unsure of what you mean sorry. The changes requested by the Recruit product team have meant changes that do not sit well within the design of the page. Then I have added more of a mess by adding conditional code like above to control the bits of content they want displayed.

Would placing this particular fragment into a partial view be more acceptable for now?

File: src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/Dtos/AccountLegalEntityDto.cs (hunk 1 above)
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/Dtos/AccountLegalEntityDto.cs]: 👍 will remove

File: src/SFA.DAS.ProviderRelationships/Services/DasRecruitService.cs (hunk 2 above)
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships/Services/DasRecruitService.cs]: 👍

File: src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/GetAccountProviderLegalEntityQueryHandler.cs (hunk 3 above)
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships/Application/Queries/GetAccountProviderLegalEntity/GetAccountProviderLegalEntityQueryHandler.cs]: Yes something has had to be descoped for our more immediate release. This is the thing to do with the ALE Id which Recruit doesn't use. We are planning to correct this soon after the Provider Recruit private beta. I will remove this method.

File: src/SFA.DAS.ProviderRelationships.Web/Views/AccountProviderLegalEntities/Get.cshtml (hunk 7 above)
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.ProviderRelationships.Web/Views/AccountProviderLegalEntities/Get.cshtml]: The behavior of not being allowed to grant permission for the recruitment operation if the provider is blocked should ideally be enforced by logic within the domain model. Currently this behavior is enforced in the view and not the model.

File: src/SFA.DAS.ProviderRelationships.Web/Controllers/AccountProvidersController.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (elliotritchie) [reply src/SFA.DAS.ProviderRelationships.Web/Controllers/AccountProvidersController.cs]: Yeah that's fine :+1:

File: src/SFA.DAS.ProviderRelationships.Web/Views/AccountProviderLegalEntities/Get.cshtml (hunk 7 above)
[the remark this answers is not in the document]
Comment (sachpatel) [reply src/SFA.DAS.ProviderRelationships.Web/Views/AccountProviderLegalEntities/Get.cshtml]: I understand the concern. I am going to leave as is. It's not ideal but this needs to go out with the logic to display the bits of content with the conditional here. I have the call to determine whether the provider is blocked in recruit happening within the QueryHandler. This is something we are going to return to very soon.
